### DISCUSSION TITLE
Schema does not match error when upgrading to Drift 2

### DISCUSSION THREAD

**Discussion by cal-g** [simolus3/drift#2211] [Q&A] [answered]
Hi,
I upgraded Drift in my project from 1.7 to the latest and get an error when `db.validateDatabaseSchema()` is called.
```
Schema does not match
recent_projects:
columns:
  time_created:
   Not equal: `NULL DEFAULT (CAST(strftime('%s', CURRENT_TIMESTAMP) AS INTEGER))` (expected) and `NULL DEFAULT (strftime('%s', CURRENT_TIMESTAMP))` (actual)
```

only for columns defined as such:
![Screenshot 2022-12-15 at 10 53 46 AM](https://user-images.githubusercontent.com/98470469/207906888-6e248b36-d5a6-4b54-9c39-e7f8771c00a1.png)

No table changes were made during the upgrade process so I wouldn't think upgrading the schema number is required. 
Do you know where I must have gone wrong ?

Thanks :)

**Comment by simolus3** [comment, marked as answer]
Good catch, thanks for bringing it up. When we changed drift to allow date times to be stored as text, we also changed the `currentDateAndTime` implementation to use `strftime` with a cast to be explicit about storing dates. Unfortunately, this now breaks existing databases because  (while practically equivalent), the two forms have different syntax and so the diff algorithm complains about it.
We could treat this case specially when comparing schemas, but I'm not sure it's worth the effort and potential confusion.

So I think there are two ways to approach this:

1. You could add a (effectively pointless) schema bump and a `TableMigration` re-creating the table - this will bring the table on the database in sync with what drift expects a `currentDateAndTime` expression to look like now.
2. You could use an alternative to `currentDateAndTime` that explicitly uses the old format. You can use the following definition instead of drift's `currentDateAndTime`. This doesn't change your database (which was created with a syntactically equivalent form of `currentDateAndTime`), but then drift really expects a `currentDateAndTime` to look the way it currently looks like in your database.

```dart
const Expression<DateTime> currentDateAndTime = _MyCurrentDateAndTimeExpression();

class _MyCurrentDateAndTimeExpression extends Expression<DateTime> {
  const _MyCurrentDateAndTimeExpression();

  @override
  void writeInto(GenerationContext context) {
    if (context.typeMapping.storeDateTimesAsText) {
      // No need to cast to a unix timestamp
      context.buffer.write('CURRENT_TIMESTAMP');
    } else {
      context.buffer.write("strftime('%s', CURRENT_TIMESTAMP)");
    }
  }
}
```

I've thought about reversing the change in drift, but that then would break apps installed with drift 2 and I think the new form really is better. So I think I'll just be mean about it and say that this is okay since it's happened in a major version update. Sorry for the annoyance with that though, I hope it can be resolved with approach 1 or 2. If you have more questions about it, please let me know!

**Reply by cal-g** [reply]
Thanks, the first approach works well.
